graph/src/data/store/mod.rs

```diff
@@ -846,6 +847,31 @@ impl Entity {
             }
         }
 
+        fn validate_fields<'a>(
+            field_names_iter: impl Iterator<Item = &'a str>,
+            object_type: &ObjectType<'_, String>,
+        ) -> Result<(), Error> {
+            let fields_set: HashSet<_> =
+                object_type.fields.iter().map(|f| f.name.as_str()).collect();
```

Comment:
I think this might be overkill for most entities which are probably small (a bit arbitrary) but this function seems to be called potentially twice per `set` operation on the entity cache . I think it may be better to either consider storing the fields on a hashset to begin with at the entity level, or ensure they are ordered by name and just lookup on the vec which *should* be a lot cheaper for smaller entities.

Reply:
cc @lutter for an opinion

Reply:
it may be a more complicated change but possibly sorting entries in the Object from intern.rs could be useful for this since we could have `field_names_iter` sorted which means that looking up is amortized and probably cheaper than creating a HashSet + a Vec for each validation

Reply:
nitpick: The naming here could also be improved. Something like `set_fields(_iter)` instead of `field_names_iter` and `object_schema` rather than `object_type`would feel a bit easier IMO but obviously it's subjective and nitpicky

Reply:
I am not sure how much of an issue this will be in the end, but I am also a little concerned about the impact of creating a `HashSet` all the time .. I think using `.contains()` will be preferable since object types generally have a small number of fields and so its likely that a quadratic algorithm here won't hurt much.

Another option would be to precompute the `HashSet` when the `InputSchema` is created, store them in a map `type name -> HashSet` and use those for this check.

A third option is to store a `Vec` of atoms for each object type in the `InputSchema` and compare against that - since you are only comparing integers at that point, and the `Vec` will fit into a cache line in almost all cases, that's probably the fastest way to do that.

Reply:
How would option 3 work @lutter?
add a new Map<ObjectType, Atom> and store the Atom corresponding to the fields.
Then wouldn't we need to look up the Atom from the pool first with the `field_name` only then we could compare right? 

I'd go with 1 for now and make the changes to use the `.contains()` let me know if 3 would be worth it. I'll do that if so.

Reply:
yeah, I think going with (1) will be fine for now; what I had in mind for (3) was a `Map<ObjectType, Vec<Atom>>` where the `Vec<Atom>` contains all the field names for the object type.

Reply:
One other small complication: the object type will also have the names of derived fields, even though they can't be set directly. Maybe we should precompute that map for (3) and only include non-derived fields for each type.